src/server/scripts/Pandaria/zone_the_wandering_isle.cpp

```diff
@@ -0,0 +1,2356 @@
+/*
+ * Copyright (C) 2008-2019 TrinityCore <http://www.trinitycore.org/>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AreaTriggerAI.h"
+#include "DB2Stores.h"
+#include "GameObject.h"
+#include "GridNotifiers.h"
+#include "Map.h"
+#include "PhasingHandler.h"
+#include "Player.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "SpellScript.h"
+
+enum CaveOfMeditationSpells
+{
+    SPELL_MEDITATION_TIMER_BAR  = 116421,
+    QUEST_THE_WAY_OF_THE_TUSHUI = 29414
+};
+
+class at_cave_of_meditation : public AreaTriggerScript
+{
+public:
+    at_cave_of_meditation() : AreaTriggerScript("at_cave_of_meditation") { }
+
+    bool OnTrigger(Player* player, AreaTriggerEntry const* /*areaTrigger*/, bool entered) override
+    {
+        if (player->IsAlive() && player->GetQuestStatus(QUEST_THE_WAY_OF_THE_TUSHUI) == QUEST_STATUS_INCOMPLETE)
+        {
+            if (entered && !player->HasAura(SPELL_MEDITATION_TIMER_BAR))
+                player->CastSpell(player, SPELL_MEDITATION_TIMER_BAR, true);
+            else
+                player->RemoveAura(SPELL_MEDITATION_TIMER_BAR);
+
+            return true;
+        }
+        return false;
+    }
+};
+
+class spell_summon_troublemaker : public SpellScript
+{
+    PrepareSpellScript(spell_summon_troublemaker);
+
+    void HandleSummon(SpellEffIndex effIndex)
+    {
+        PreventHitDefaultEffect(effIndex);
+        uint32 entry = GetSpellInfo()->GetEffect(effIndex)->MiscValue;
+        SummonPropertiesEntry const* properties = sSummonPropertiesStore.LookupEntry(GetSpellInfo()->GetEffect(effIndex)->MiscValueB);
+        int32 duration = GetSpellInfo()->GetDuration();
+
+        if (!entry || !properties || !duration)
+            return;
+
+        int32 radius;
+        if (urand(0, 2) == 0)
```

Comment:
So one of three chances to be between 0 to 6 yards but otherwise 7? Where does that come from?

Also, my scripting is rusty, but pretty sure we have utilities for randomizing angles and calculating positions based off an origin and an angle - use those instead of hardcoding magic values.

Reply:
Actually this is a typo, it should be lower chance of radius 0 than 1-7. If the radius is 0, the angle doesn't matter anymore, and any random angle selected will still be in the middle.
This way I want to ensure that there won't be a lot more spawns on position 0, 0 then on the others.

This is how the spawn positions look like (every dot is possible spawn position):
![image](https://user-images.githubusercontent.com/32191/62536987-a748bd00-b84f-11e9-99e2-493d4ed7dbc3.png)

Are those utilities still applicable for such situation? Is it really necessary to use them if there are any?
It would be pain in the ass for me finding them, understanding how they work and applying them :cold_sweat: (is there anyone who can help me with this?)

Reply:
```
Position pos = { 1181.75f, 3444.5f, 102.9385f, 0.0f };
pos.SetOrientation(<angle randomization, don't need to normalize>);
pos.RelocateOffset(<radius randomization>);
```

Something along those lines